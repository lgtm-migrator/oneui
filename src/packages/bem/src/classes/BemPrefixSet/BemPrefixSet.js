/**
 * @type {Array.<string>}
 */
const ALLOWED_PREFIX_NAMES = ['elem', 'mod', 'value'];

/**
 * @type {RegExp}
 */
const PREFIX_PATTERN = /[_-]+/;

/**
 * @type {string}
 */
const DEFAULT_ELEM_PREFIX = '__';

/**
 * @type {string}
 */
const DEFAULT_MOD_PREFIX = '--';

/**
 * @type {string}
 */
const DEFAULT_VALUE_PREFIX = '_';

/**
 * @type {Object.<string, string>}
 */
const defaultPrefixes = {
    elem: DEFAULT_ELEM_PREFIX,
    mod: DEFAULT_MOD_PREFIX,
    value: DEFAULT_VALUE_PREFIX
}

export class BemPrefixSetError extends Error {}

export default class BemPrefixSet {
    constructor(prefixes = {}) {
        const { elem, mod, value } = {
            ...defaultPrefixes,
            ...prefixes
        };
        BemPrefixSet.validate({ elem, mod, value });
        Object.assign(this, { elem, mod, value });
    }

    /**
     * @param {BemPrefixSet} prefixSet
     * @throws {BemPrefixSetError}
     */
    static validate(prefixSet) {
        Object.keys(prefixSet).forEach(prefixName => {
            // Validating keys name
            if (ALLOWED_PREFIX_NAMES.includes[prefixName] === false) {
                throw new BemPrefixSetError(
                    `BEM prefixes declaration includes unknown property '${prefixName}'. ` +
                        `Allowed values are ${ALLOWED_PREFIX_NAMES.join(', ')}.`
                );
            }

            // Validating values
            const prefixValue = prefixSet[prefixName];
            const prefixValueMatch = prefixValue.match(PREFIX_PATTERN);
            if (prefixValueMatch === null) {
                throw new BemPrefixSetError(
                    `BEM ${prefixName} has invalid value: '${prefixValue}'. It contains invalid characters or it is empty.`
                );
            }
        });
    }
}
