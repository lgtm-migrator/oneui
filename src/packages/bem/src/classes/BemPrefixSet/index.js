export { default, BemPrefixSetError } from './BemPrefixSet';
