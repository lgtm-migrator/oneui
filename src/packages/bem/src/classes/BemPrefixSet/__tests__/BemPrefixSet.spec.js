import BemPrefixSet, { BemPrefixSetError } from '..';

describe('BemPrefixSet', () => {

    describe('module export', () => {
        it('should export BemPrefixSet as default', () => {
            expect(BemPrefixSet).toBeTruthy();
        })
        it('should export BemPrefixSetError as named export', () => {
            expect(BemPrefixSetError).toBeTruthy();
        })
    });

    describe('#constructor', () => {
        it('should create set of BEM prefixes', () => {
            const prefixes = {
                elem: '__',
                mod: '--',
                value: '_'
            };
            const bemPrefixSet = new BemPrefixSet(prefixes);
            expect(bemPrefixSet).toEqual(prefixes);
        });
        it('should create a default set of BEM prefixes if nothing is passed', () => {
            const bemPrefixSet = new BemPrefixSet();
            expect(bemPrefixSet).toEqual({
                elem: '__',
                mod: '--',
                value: '_'
            });
        });
        it('should set "elem" to default if it is not set', () => {
            const prefixes = {
                mod: '___',
                value: '----'
            };
            const bemPrefixSet = new BemPrefixSet(prefixes);
            expect(bemPrefixSet).toEqual({
                ...prefixes,
                elem: '__',
            });
        });
        it('should set "mod" to default if it is not set', () => {
            const prefixes = {
                elem: '___',
                value: '____'
            };
            const bemPrefixSet = new BemPrefixSet(prefixes);
            expect(bemPrefixSet).toEqual({
                ...prefixes,
                mod: '--',
            });
        });
        it('should set "value" to default if it is not set', () => {
            const prefixes = {
                elem: '____',
                mod: '---'
            };
            const bemPrefixSet = new BemPrefixSet(prefixes);
            expect(bemPrefixSet).toEqual({
                ...prefixes,
                value: '_',
            });
        });
        it('should throw if "elem" prefix is invalid', () => {
            const prefixes = {
                elem: '~',
                mod: '--',
                value: '_'
            };
            expect(() => new BemPrefixSet(prefixes)).toThrowErrorMatchingSnapshot();
        });
        it('should throw if "mod" prefix is invalid', () => {
            const prefixes = {
                elem: '__',
                mod: '~',
                value: '_'
            };
            expect(() => new BemPrefixSet(prefixes)).toThrowErrorMatchingSnapshot();
        });
        it('should throw if "value" prefix is invalid', () => {
            const prefixes = {
                elem: '__',
                mod: '--',
                value: '~'
            };
            expect(() => new BemPrefixSet(prefixes)).toThrowErrorMatchingSnapshot();
        });
        it('should throw if "elem" prefix is empty', () => {
            const prefixes = {
                elem: '',
                mod: '--',
                value: '_'
            };
            expect(() => new BemPrefixSet(prefixes)).toThrowErrorMatchingSnapshot();
        });
        it('should throw if "mod" prefix is empty', () => {
            const prefixes = {
                elem: '__',
                mod: '',
                value: '_'
            };
            expect(() => new BemPrefixSet(prefixes)).toThrowErrorMatchingSnapshot();
        });
        it('should throw if "value" prefix is empty', () => {
            const prefixes = {
                elem: '__',
                mod: '--',
                value: ''
            };
            expect(() => new BemPrefixSet(prefixes)).toThrowErrorMatchingSnapshot();
        });
    });
});
