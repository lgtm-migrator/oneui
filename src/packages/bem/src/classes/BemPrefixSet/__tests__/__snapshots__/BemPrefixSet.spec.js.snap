// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BemPrefixSet #constructor should throw if "elem" prefix is empty 1`] = `"BEM elem has invalid value: ''. It contains invalid characters or it is empty."`;

exports[`BemPrefixSet #constructor should throw if "elem" prefix is invalid 1`] = `"BEM elem has invalid value: '~'. It contains invalid characters or it is empty."`;

exports[`BemPrefixSet #constructor should throw if "mod" prefix is empty 1`] = `"BEM mod has invalid value: ''. It contains invalid characters or it is empty."`;

exports[`BemPrefixSet #constructor should throw if "mod" prefix is invalid 1`] = `"BEM mod has invalid value: '~'. It contains invalid characters or it is empty."`;

exports[`BemPrefixSet #constructor should throw if "value" prefix is empty 1`] = `"BEM value has invalid value: ''. It contains invalid characters or it is empty."`;

exports[`BemPrefixSet #constructor should throw if "value" prefix is invalid 1`] = `"BEM value has invalid value: '~'. It contains invalid characters or it is empty."`;
